actors/miner/src/notifications.rs

```diff
@@ -0,0 +1,182 @@
+use crate::{
+    PieceActivationManifest, PieceChange, SectorChanges, SectorContentChangedParams,
+    SectorContentChangedReturn, SECTOR_CONTENT_CHANGED,
+};
+use fil_actors_runtime::runtime::Runtime;
+use fil_actors_runtime::{
+    actor_error, ActorError, AsActorError, SendError, STORAGE_MARKET_ACTOR_ADDR,
+};
+use fvm_ipld_encoding::ipld_block::IpldBlock;
+
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::sector::SectorNumber;
+use fvm_shared::sys::SendFlags;
+use num_traits::Zero;
+use std::collections::BTreeMap;
+
+pub struct ActivationNotifications<'a> {
+    pub sector_number: SectorNumber,
+    pub sector_expiration: ChainEpoch,
+    pub pieces: &'a Vec<PieceActivationManifest>,
+}
+
+/// Sends notifications of sector and piece activation to nominated receiving actors.
+/// Inputs are per-sector, matching parameters provided by external callers.
+/// This method groups inputs by receiving actor and sends a single notification per actor.
+/// Notifications are fire-and-forget, so there is no return value.
+/// If require_success is true, this method will return an error if any of the notifications fail
+/// or are rejected by the receiving actor.
+pub fn notify_data_consumers(
+    rt: &impl Runtime,
+    activations: &[ActivationNotifications],
+    require_success: bool,
+) -> Result<(), ActorError> {
+    // Inputs are grouped by sector -> piece -> notifee+payload
+    // Regroup by notifee -> sector -> piece+payload for sending.
+    let mut activations_by_notifee =
+        BTreeMap::<Address, BTreeMap<SectorNumber, Vec<PieceChange>>>::new();
+    // Collect each sector's expiration for sending to each receiver of a notification.
+    let mut sector_expirations = BTreeMap::<SectorNumber, ChainEpoch>::new();
```

Comment:
I will change this because it's also more natural. I would have done this in the first place if not concerned about the efficiency loss of iterating the input twice. The difference in the BTreeMap building is surely only a small constant anyway.

Rant: Having to still think about this kind of thing is one of the great failures of modern programming languages. It shouldn't matter.